examples/opensora_cai/opensora/models/layers/rotary_embedding.py

```diff
@@ -0,0 +1,149 @@
+"""
+Source: https://github.com/lucidrains/rotary-embedding-torch/
+"""
+from math import pi
+from typing import Optional
+
+try:
+    from typing import Literal
+except ImportError:
+    from typing_extensions import Literal  # FIXME: python 3.7
+
+import numpy as np
+
+from mindspore import Parameter, Tensor, dtype, nn, ops
+
+
+def rotate_half(x: Tensor) -> Tensor:
```

Comment:
existed in mindone/models/modules/pos_embed.py, maybe reuse

Reply:
The rotary embedding in `mindone/models/modules/pos_embed.py` is slightly different from the one used is OpenSora (it uses [this package](https://github.com/lucidrains/rotary-embedding-torch)). I think it's better to preserve the imported `RotaryEmbedding` class intact for now. Let me see later if I can merge them.